### PR TITLE
Adding retries for **ANY** http request.

### PR DESCRIPTION
**Pull request by dhermes**
Will always retry on 429, 500, 502, 503 and 504. Based on:
https://cloud.google.com/storage/docs/json_api/v1/how-tos/resumable-upload#handling_errors

### CODE REVIEW

File: google/resumable_media/_helpers.py [hunk 1]
```diff
@@ -131,4 +183,19 @@ def http_request(transport, method, url, data=None, headers=None):
...
         object: The return value of ``transport.request()``.
     """
-    return transport.request(method, url, data=data, headers=headers)
+    response = transport.request(method, url, data=data, headers=headers)
+    if get_status_code(response) not in RETRYABLE:
+        return response
+
+    total_sleep = 0.0
+    num_retries = 0
+    while total_sleep < MAX_CUMULATIVE_RETRY:
```
Comment (jonparrott) [inline google/resumable_media/_helpers.py]: can you pull this behavior into a separate method/decorator/whatever?

Comment (dhermes) [inline google/resumable_media/_helpers.py]: Not sure what you have in mind? Like a `do_stuff(func)` wrapper?

Comment (jonparrott) [inline google/resumable_media/_helpers.py]: it seems like this retry logic is "unrelated" to the request object, so I'd prefer to pull it into its own thing, maybe something (conceptually) like:

```python
func = functools.partial(transport.request, method, url, data=data, headers)
predicate = lambda response: get_status_code(response) not in RETRYABLE
return _retry(func, predicate)
``

**Comment by dhermes** [comment]
@jonparrott Requested re-factor has been done. PTAL

File: google/resumable_media/_helpers.py [hunk 2]
```diff
@@ -114,6 +140,81 @@ def require_status_code(response, status_codes, callback=do_nothing):
...
+    Success is determined by ``predicate(result)`` being true.
+
+    Will retry until :attr:`MAX_CUMULATIVE_RETRY` seconds of wait time
+    have accrued. Uses :func:`calculate_retry_wait` to double the
+    wait time (with jitter) after each attempt.
+
+    Args:
+        func (Callable): A callable that takes no arguments and produces
+            a result which will be checked by ``predicate``.
+        predicate (Callable): Determines if the result is valid.
```
Comment (jonparrott) [inline google/resumable_media/_helpers.py]: Callable[Any, bool]

File: google/resumable_media/_helpers.py [hunk 3]
```diff
@@ -114,6 +140,81 @@ def require_status_code(response, status_codes, callback=do_nothing):
...
+            return result
+
+    return result
+
+
+def not_retryable_predicate(response):
+    """Determines if a ``response`` is retryable.
+
+    Args:
+        object: The return value of ``transport.request()``.
```
Comment (jonparrott) [inline google/resumable_media/_helpers.py]: This is always `google.auth.transport.Response`.

File: google/resumable_media/_helpers.py (hunk 2 above)
Comment (dhermes) [inline google/resumable_media/_helpers.py]: Done (and folded into the re-factor commit)

File: google/resumable_media/_helpers.py (hunk 3 above)
Comment (dhermes) [inline google/resumable_media/_helpers.py]: The idea is that `transport` can be anything, which is why I'm strongly considering making `google.resumable_media.FOO` subpackages so it is **ALWAYS** explicit

**Pull request merged**